sklearn/_fast_gradient_boosting/_predictor.pyx

```diff
@@ -0,0 +1,100 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: language_level=3
+
+# Author: Nicolas Hug
+
+cimport cython
+from cython.parallel import prange
+import numpy as np
+cimport numpy as np
+
+from .types cimport X_DTYPE_C
+from .types cimport Y_DTYPE_C
+from .types cimport X_BINNED_DTYPE_C
+
+
+cdef packed struct node_struct:
+    # Equivalent struct to PREDICTOR_RECORD_DTYPE to use in memory views. It
+    # needs to be packed since by default numpy dtypes aren't aligned
+    Y_DTYPE_C value
+    unsigned int count
+    unsigned int feature_idx
+    X_DTYPE_C threshold
+    unsigned int left
+    unsigned int right
+    Y_DTYPE_C gain
+    unsigned int depth
+    unsigned char is_leaf
+    X_BINNED_DTYPE_C bin_threshold
+
+
+def _predict_from_numeric_data(nodes, numeric_data, out):
+    _predict_from_numeric_data_parallel(nodes, numeric_data, out)
+
+
+def _predict_from_binned_data(nodes, binned_data, out):
+    _predict_from_binned_data_parallel(nodes, binned_data, out)
+
+
+cdef void _predict_from_numeric_data_parallel(
+        node_struct [:] nodes,
+        const X_DTYPE_C [:, :] numeric_data,
+        Y_DTYPE_C [:] out):
+
+    cdef:
+        int i
+
+    for i in prange(numeric_data.shape[0], schedule='static', nogil=True):
+        out[i] = _predict_one_from_numeric_data(nodes, numeric_data, i)
+
+
+cdef inline Y_DTYPE_C _predict_one_from_numeric_data(
+        node_struct [:] nodes,
+        const X_DTYPE_C [:, :] numeric_data,
+        const int row) nogil:
+    # Need to pass the whole array and the row index, else prange won't work.
+    # See issue Cython #2798
+
+    cdef:
+        node_struct node = nodes[0]
+
+    while True:
+        if node.is_leaf:
```

Comment:
isn't `is_leaf` derivable from `right` and/or `left`? If yes, wouldn't it be 1) saving memory 2) be safer (i.e. forgetting to set the flag, hence some invalid memory access issue) to remove it from the struct?

Reply:
Yes an implicit way of determining wether a node is a leaf is that `left == right == 0`.

I personally find the explicit way (`is_leaf` attribute) better:
- it's explicit
- memory usage is very minimal (1 byte per node)

I don't see any way of forgetting to set the flag: It's only set in the grower in `_fill_predictor_node_array`.

Reply:
I'm more concerned about a malicious user giving invalid input to this function and as a result access a part of memory they shouldn't. But I guess we generally don't worry about that here I suppose.

Reply:
Although when it comes to maintainability of the code, future developers may somehow forget to set it correctly?

Reply:
>I'm more concerned about a malicious user giving invalid input to this function and as a result access a part of memory they shouldn't. 

That'd be a concern regardless of the convention that we choose for defining a leaf right?

>future developers may somehow forget to set it correctly?

It would (should) be immediately caught with the tests? I find it easier for future developers to be aware of the `is_leaf` field, than to be aware of the implicit convention that if a node's childs are both null then it's a leaf. Honestly that's exactly the kind of thing that I personally found confusing when reading the existing tree code.